### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by pwojnaro**
This PR, when merged, will introduce the initial implementation of `@planship/svelte` library. The `@planship/svelte` is based on `@planship/fetch` and it uses Svlte context and stores to simplify consuming Planship data and functionality in Svelte and SvelteKit apps.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
+# planship-svelte
+
+Welcome to `@planship/svelte`, library that enables entitlements, metering, plan packaging, and customer/subscription management in your [Svelte](https://svelte.dev/) and [SvelteKit]https://kit.svelte.dev/) apps powered by [Planship](https://planship.io). This SDK is built on top of the [@planship/fetch](https://github.com/planship/planship-js/tree/master/packages/fetch) JavaScript library and it uses the Svelte [Context API](https://learn.svelte.dev/tutorial/context-api) and [stores](https://svelte.dev/docs/svelte-store) under the good.
```
Comment (trentpeterson) [inline README.md]: 1. , library -> , a Svelte library
2. and it uses -> and uses
3. under the good -> under the hood? (might consider rewording this)

File: README.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+npm install @planship/svelte
+# or
+yarn add @planship/svelte
+# or
+pnpm add @planship/svelte
+```
+
+## Working with entitlements and other customer data - `PlanshipCustomerProvider` and `usePlanshipCustomer`
+
+In most rendering scenarios, your app will need to fetch and evaluate Planship entitlements for a specific customer. This can be accomplished with `PlanshipCustomerProvider` which initializes a Planship API instance for a specific customer, continously fetches their `entitlements`, and exposes them via a reactive Svelte store.
```
Comment (trentpeterson) [inline README.md]: which -> , which

File: README.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+
+### Initialization
+
+First, initialize `PlanshipCustomerProvider` near the root of your app. Please note, that `PlanshipCustomerProvider` can be used only within a context where the Planship customer ID (typically your current user) is known (E.g. a top level `layout.svelte` for a group of routes that require authentication).
```
Comment (trentpeterson) [inline README.md]: Please note, -> Please note

File: README.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+
+```svelte
+<script lang="ts">
+  import { PlanshipCustomerProvider } from `@planship/svelte`;
+  import { getContext } from 'svelte';
+
+  // Planship access token getter function that retrives a Planship access token from the application backend
+  const getAccessToken = () => {
+    return fetch("/api/planshipToken").then((r: Response) => r.text())
```
Comment (trentpeterson) [inline README.md]: single quotes

File: README.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+First, initialize `PlanshipCustomerProvider` near the root of your app. Please note, that `PlanshipCustomerProvider` can be used only within a context where the Planship customer ID (typically your current user) is known (E.g. a top level `layout.svelte` for a group of routes that require authentication).
+
+
+```svelte
+<script lang="ts">
+  import { PlanshipCustomerProvider } from `@planship/svelte`;
+  import { getContext } from 'svelte';
+
+  // Planship access token getter function that retrives a Planship access token from the application backend
```
Comment (trentpeterson) [inline README.md]: retrives -> retrieves

File: README.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+  const getAccessToken = () => {
+    return fetch("/api/planshipToken").then((r: Response) => r.text())
+  }
+
+  // retrieve currently authanticated user stored in the context
```
Comment (trentpeterson) [inline README.md]: authanticated -> authenticated

File: README.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+  const user: Readable = getContext('$currentUser')
+
```
Comment (trentpeterson) [inline README.md]: Remove newline and indent `script` block contents.

File: README.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+</script>
+
+<PlanshipCustomerProvider config={{ slug:'clicker-demo', customerId: $user.id, getAccessToken }}>
+  <slot/>
+</PlanshipCustomerProvider>
+```
+
+`PlanshipCustomerProvider` has to be configured via `config` parameter that has the following, mandatory properties:
```
Comment (trentpeterson) [inline README.md]: following, -> following

File: README.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+- **`slug`** - your [Planship product slug](https://docs.planship.io/concepts/products/)
+- **`customerId`** - your [Planship customer id](https://docs.planship.io/concepts/customers/)
+- **`getAccessToken`** - function (or promise) that returns a Planship access token
+
+### Authentication with Planship and `getAccessToken` function
+
+Since `PlanshipCustomerProvider` is initialized in the client-side code (or in both client- and server-side code when used with SvelteKit), security credentials cannot be used to authenticate with Planship. Instead, retrieve a Planship access token in your application server, and return it using an existing, secure connection to your backend.
```
Comment (trentpeterson) [inline README.md]: 1. in the client-side -> on the client-side
2. in yoru application server -> on your application server
3. existing, -> existing

File: README.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+  const token = await planship.getAccessToken()
+  return new Response(token.accessToken);
+}
+```
+
+### Consuming entitlements in your pages and components
+
+With `PlanshipCustomerProvider` intialized,  call `usePlanshipCustomer` in any page or component code to access customer entitlements.
+
+The example below shows how customer entitlements are retrieved, and a simple boolean feature called `advanced-analytics` is used to conditionally render an `<a>` element inside a Svelte component. Please note that `entitlements` dictionary is wrapped in a Svelte store.
```
Comment (trentpeterson) [inline README.md]: 1. Suggest rewording first sentence: The example below shows how to retrieve customer entitlements and use them for conditional rendering within your pages and components.
2. `entitlements` dictionary -> the `entitlements` dictionary

File: README.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+```svelte
+<script lang='ts'>
+  import { usePlanshipCustomer } from '@planship/svelte';
```
Comment (trentpeterson) [inline README.md]: Be consistent with semicolons

File: README.md [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+  const { entitlements } = usePlanshipCustomer()
+</script>
+
+{#if $entitlements['advanced-analytics']}
+  <a to="/analytics">
+    Analytics
```
Comment (trentpeterson) [inline README.md]: I'd move this to a single line

File: README.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+  </a>
+{/if}
+```
+
+### Fetching additional customer data from Planship
+
+Your app may need to fetch additional customer data from Planship (E.g. customer subscription or usage data). To accomplish any Planship API operation use an instance of the [Planship Customer API client](https://github.com/planship/planship-js/blob/master/packages/fetch/docs/interfaces/PlanshipCustomerApi.md) returned by the `usePlanshipCustomer` function.
+
+Below is an example Svelte script that retrieves a list of subscriptions for the current customer.
```
Comment (trentpeterson) [inline README.md]: Svelte script? Svelte component?

File: README.md [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+{#if $entitlements.advancedAnalytics}
+  <a to="/analytics">
+    Analytics
+  </a>
+{/if}
+```
+
+## Working with plans and other product data - `PlanshipProvider` and `usePlanship`
+
+If the current customer context is unknown (E.g. pages of your app that don't require authentication like a pricing page) , use `PlanshipProvider` that provides a Svelte context that contains an instance of the [Planship API client](https://github.com/planship/planship-js/blob/master/packages/fetch/docs/interfaces/PlanshipApi.md). `PlanshipProvider` is not customer specific, so you need only your [Planship product slug](https://docs.planship.io/concepts/products/) and a [Planship token getter](#authentication-with-planship-and-getaccesstoken-function) to initialize it.
```
Comment (trentpeterson) [inline README.md:140]: Suggested rewording:
> If the current customer context is unknown (E.g. on pages of your app that don't require authentication like a pricing page) , `PlanshipProvider` provides a Svelte context containing an instance of the [Planship API client](https://github.com/planship/planship-js/blob/master/packages/fetch/docs/interfaces/PlanshipApi.md). `PlanshipProvider` is not customer specific, so you only need your [Planship product slug](https://docs.planship.io/concepts/products/) and a [Planship token getter](#authentication-with-planship-and-getaccesstoken-function) to initialize it.

File: package.json [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    "eslint-plugin-svelte": "^2.40.0",
+    "svelte": "^4.2.17",
+    "svelte-check": "^3.7.1",
+    "svelte-eslint-parser": "^0.39.1"
+  },
+  "peerDependencies": {
+    "svelte": "^3.54.0 || ^4.0.0 || ^5.0.0-next.0"
+  },
+  "dependencies": {
+    "@planship/fetch": "0.3.0-alpha6"
```
Comment (trentpeterson) [inline package.json]: Move to devDeps

File: src/PlanshipCustomerProvider.svelte [hunk 16]
```diff
@@ -0,0 +1,36 @@
+<script lang="ts">
+    import { PlanshipCustomer } from '@planship/fetch';
```
Comment (trentpeterson) [inline src/PlanshipCustomerProvider.svelte:2]: Be consistent with semicolons

File: src/PlanshipCustomerProvider.svelte [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    import type { CustomerProviderConfig } from './types.js';
+    import { writable } from 'svelte/store';
+
+    export let config: CustomerProviderConfig
+
+    const entitlements = writable<Entitlements>({})
+    const subscriptions = writable<CustomerSubscriptionWithPlan[]>([])
+
+    const { baseUrl, webSocketUrl, slug, getAccessToken, customerId } = config
+
```
Comment (trentpeterson) [inline src/PlanshipCustomerProvider.svelte]: Remove extra newlines

File: src/PlanshipProvider.svelte [hunk 18]
```diff
@@ -0,0 +1,22 @@
+<script lang="ts">
+  import { Planship } from '@planship/fetch';
+  import { setPlanshipContext } from './context.js';
+  import type { ProviderConfig } from './types.js';
+
+  export let config: ProviderConfig
+
+  const { baseUrl, webSocketUrl, slug, getAccessToken } = config
+
+
```
Comment (trentpeterson) [inline src/PlanshipProvider.svelte]: Remove extra newlines

File: src/usePlanship.ts [hunk 19]
```diff
@@ -0,0 +1,6 @@
+import { getPlanshipContext, IPlanshipContext } from './context.js'
+
+export function usePlanship(context?: IPlanshipContext): IPlanshipContext {
+  if (context) return context
```
Comment (trentpeterson) [inline src/usePlanship.ts]: `return context || getPlanshipContext()` ?

File: src/usePlanshipCustomer.ts [hunk 20]
```diff
@@ -0,0 +1,26 @@
+import { CustomerSubscriptionWithPlan, Entitlements, PlanshipCustomerApi } from '@planship/fetch'
+import { EntitlementsBase } from './types.js'
```
Comment (trentpeterson) [inline src/usePlanshipCustomer.ts]: Import types? I know we talked about whether this is required but I guess we should be consistent.

File: src/usePlanshipCustomer.ts [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+import { getPlanshipCustomerContext, IPlanshipCustomerContext } from './customerContext.js'
+import { get, Readable, derived } from 'svelte/store'
+
+
+interface ICustomerContext<T extends EntitlementsBase> {
+    planshipCustomerApiClient?: PlanshipCustomerApi
+    entitlements: Readable<T>
+    subscriptions: Readable<CustomerSubscriptionWithPlan[]>
+  }
```
Comment (trentpeterson) [inline src/usePlanshipCustomer.ts]: Unindent this block two spaces

File: src/customerContext.ts [hunk 22]
```diff
@@ -0,0 +1,28 @@
+import { CustomerSubscriptionWithPlan, Entitlements, PlanshipCustomerApi } from '@planship/fetch'
+import { getContext, setContext } from 'svelte'
+import { Readable } from 'svelte/store'
+
+const _contextKey = '$$_planshipCustomer'
+
+export interface IPlanshipCustomerContext {
+    planshipCustomerApiClient?: PlanshipCustomerApi
```
Comment (trentpeterson) [inline src/customerContext.ts]: Unindent this block two spaces

**Review by trentpeterson — CHANGES_REQUESTED** [review]
How would you use Planship functionality in a page or layout .ts file? Is it possible to do so (and do so safely)?

General comment about being consistent with styling (semicolons, newlines, braces around conditional code, etc). Definitely need to agree on a standard and then use it in all projects.

File: README.md (hunk 7 above)
Comment (pwojnaro) [reply README.md]: The script block contents are already indented (2 spaces)? I removed most newlines, but left the ones before comments, more readable this way.

File: README.md (hunk 9 above)
Comment (pwojnaro) [reply README.md]: Re 1: it's in the client-side **code**, so I'm assuming it should stay **in**?

File: package.json (hunk 15 above)
Comment (pwojnaro) [reply package.json]: Can do, but is there a strong reason why? It doesn't really matter where do you put it in Vite, but I like an indication that this dep will actually end up in your runtime code.

Comment (pwojnaro) [reply package.json]: Update: since this a library, moving to devdeps actually breaks it (app cannot find it anymore).

File: README.md (hunk 13 above)
Comment (pwojnaro) [reply README.md]: Changing to Svelte component initialization code. New wording:

> The example below shows Svelte component (or page) initialization code that retrieves a list of subscriptions for the current customer.

File: README.md (hunk 14 above)
Comment (pwojnaro) [reply README.md:140]: Done.

File: src/PlanshipCustomerProvider.svelte (hunk 16 above)
Comment (pwojnaro) [reply src/PlanshipCustomerProvider.svelte:2]: Done

File: src/PlanshipProvider.svelte (hunk 18 above)
Comment (pwojnaro) [reply src/PlanshipProvider.svelte]: Done

File: src/customerContext.ts (hunk 22 above)
Comment (pwojnaro) [reply src/customerContext.ts]: Done

File: src/usePlanshipCustomer.ts (hunk 21 above)
Comment (pwojnaro) [reply src/usePlanshipCustomer.ts]: Done.
